playgrounds/eav-checkbox-playground.R

```diff
@@ -0,0 +1,559 @@
+rm(list = ls(all = TRUE)) #Clear the memory of variables from previous run. This is not called by knitr, because it's above the first chunk.
+
+# ---- load-sources ------------------------------------------------------------
+
+# ---- load-packages -----------------------------------------------------------
+library("magrittr")
+requireNamespace("dplyr")
+requireNamespace("readr")
+requireNamespace("testit")
+
+# ---- declare-globals ---------------------------------------------------------
+# as projects are only containing test data, okay to commit URL and tokens
+# to Github in private repo
+# remove if we ever submit a PR to the main project
+redcap_uri <- "https://redcap.hbcd.cmig-ucsd.org/api/"
+
+# token      <- "53320377CD34F60392870F5F6C20953B" # longitudinal & repeating
+# token      <- "778FAF179F3DBEDBC1A3CA1CAB2B36EF" # repeating
+# token      <- "831D17A42D03BC448A7B1A08CDE19AD9" # longitudinal
+token      <- "B77827EB144D9B7D74F01E932C9F0B29" # neither
+
+fields  <- NULL
+# fields   <- c(
+#   # "record_id",
+#   "form_1_check_001"
+# )
+forms   <- NULL
+# forms   <- c(
+#   "form_1",
+#   # "form_2",
+#   "form_3"
+#   # "form_4"
+# )
+events  <- NULL
+# events  <- c(
+#   "event_1_arm_1"
+#   # "event_3_arm_1"
+# )
+records <- NULL
+
+blank_for_gray_form_status <- TRUE
+.complete_value_for_untouched_forms <- dplyr::if_else(
+  blank_for_gray_form_status,
+  NA_character_,
+  as.character(REDCapR::constant("form_incomplete"))
+)
+
+# NOTE: I think that an additional function parameter to specify how to handle
+#   untouched checkbox items might make sense. But we can also make it dependent
+#   on `blank_for_gray_form_status` if you prefer that. I just think some people
+#   might want to make different selections for either.
+blank_for_untouched_checkbox <- TRUE
+
+# ---- load-data ---------------------------------------------------------------
+# View(REDCapR::redcap_variables(redcap_uri, token)$data)
+system.time({
+  meta <- REDCapR:::redcap_metadata_internal(
+    redcap_uri,
+    token
+  )
+
+  ds_metadata          <- meta$d_variable
+  ds_metadata_plumbing <- ds_metadata |>
+    dplyr::filter(
+      plumbing
+    )
+
+  # only include selected fields/forms
+  if (!is.null(fields) & !is.null(forms)) {
+    ds_metadata <- ds_metadata |>
+      dplyr::filter(
+        field_name_base %in% fields | form_name %in% forms
+      )
+  } else if (!is.null(fields)) {
+    ds_metadata <- ds_metadata |>
+      dplyr::filter(
+        field_name_base %in% fields
+      )
+  } else if (!is.null(forms)) {
+    ds_metadata <- ds_metadata |>
+      dplyr::filter(
+        form_name %in% forms
+      )
+  }
+
+  col_types <- REDCapR::redcap_metadata_coltypes(
+    redcap_uri,
+    token,
+    print_col_types_to_console = FALSE
+  )
+
+  # only include selected fields/forms
+  col_types[[1]] <- col_types[[1]][ds_metadata$field_name]
+
+  ds_eav <- REDCapR:::redcap_read_eav_oneshot(
```

Comment:
I tried to google what's flat and eva in REDCap with no success. I saw them and tried them on GUI yesterday; it seems there are different ways to organize export data. Maybe could you give me some more details? BTW, a very silly question: what does `eva` mean?

Reply:
I think it's based on this data model: https://en.wikipedia.org/wiki/Entity%E2%80%93attribute%E2%80%93value_model

The difference is what you mention below, it's a longer format. While the standard export returns one row per participant (or participant/event if longitudinal) and one column per variable *independent of whether a given variable has a value for that participant*, e.g.,

```
id  var_1   var_2   var_3
ID1 1       NULL    NULL
ID2 NULL    NULL    1
```

The EAV export only returns values that actually exist

```
id  var   value
ID1 var_1 1
ID2 var_3 1
```

In terms of data storage that makes it more efficient I guess. But the main advantage for our use case is that the `eav` approach exports the checkbox items correctly. While the standard approach fills all non-chosen items with 0s (even if no checkbox option was marked or the form wasn't even administered), the EAV only displays the ones that have been selected. That's why we can use it to construct the database that we want that has

- the same shape as the standard approach (one row per participant, one column per variable)
- but correct representation of checkboxes, i.e., `NA/NULL` if no checkbox option was chosen